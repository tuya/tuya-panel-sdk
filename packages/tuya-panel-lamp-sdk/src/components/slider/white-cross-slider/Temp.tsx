/* eslint-disable @typescript-eslint/ban-ts-comment */
import React, { Component } from 'react';
import Color from 'color';
import { View, StyleSheet, Animated, Easing, StyleProp, TextStyle } from 'react-native';
import { LinearGradient, TYText } from 'tuya-panel-kit';
import { Svg, Path, Rect } from 'react-native-svg';
import ColorUtils from '../../../utils/color';

type Props = {
  /**
   * 文本样式
   */
  textStyle: StyleProp<TextStyle>;
  /**
   * 色温值
   */
  temp: number;
  /**
   * 尺寸
   */
  size: number;
  /**
   * 轨道宽度
   */
  trackWidth: number;
  /**
   * 标签文本
   */
  label: string;
  /**
   * 格式化百分比
   */
  formatPercent?: (v: number) => number;
  /**
   * 最小值
   */
  min: number;
  /**
   * 最大值
   */
  max: number;
};

export default class Temperature extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.animation.setValue(this.props.temp);
    this.animation.addListener(this.handleUpdateThumb);
  }

  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.temp !== this.props.temp) {
      this.needUpdate = true;
    }
  }

  componentDidUpdate() {
    if (this.needUpdate) {
      this.needUpdate = false;
      this.runAnimation(this.props.temp);
    }
  }

  private animation: Animated.Value = new Animated.Value(0);
  private textRef: TYText;
  private needUpdate = false;
  private pathRef: Path;

  handleUpdateThumb = ({ value }: { value: number }) => {
    const color = ColorUtils.brightKelvin2rgb(1000, value);
    const rgb = Color(color).rgbArray();
    this.pathRef &&
      this.pathRef.setNativeProps({ fill: [0, rgb[0] / 255, rgb[1] / 255, rgb[2] / 255, 1] });

    const percent = this.formatPercent(value);
    // @ts-ignore
    this.textRef && this.textRef.setText(`${this.props.label}: ${percent}%`);
  };

  runAnimation(value: number) {
    this.animation.stopAnimation();
    Animated.timing(this.animation, {
      toValue: value,
      duration: 16,
      easing: Easing.linear,
    }).start();
  }

  update(bright: number) {
    this.runAnimation(bright);
  }

  formatPercent(value: number) {
    const { min, max, formatPercent } = this.props;
    if (formatPercent) {
      return formatPercent(value);
    }

    return Math.round(((value - min) * 100) / (max - min));
  }

  render() {
    const { size, temp, label, min, max, trackWidth, textStyle = { fontSize: 12 } } = this.props;
    const color = ColorUtils.brightKelvin2rgba(1000, temp);
    const { fontSize } = StyleSheet.flatten(textStyle);
    const lineHeight = fontSize * 1.5;
    return (
      <View style={{ width: trackWidth, height: size }}>
        <View
          style={{
            width: trackWidth,
            height: size,
            borderRadius: trackWidth / 2,
            overflow: 'hidden',
          }}
        >
          <LinearGradient
            gradientId="white-cross-temp"
            x1="0%"
            x2="0%"
            y1="0%"
            y2="100%"
            stops={{
              '0%': '#CDECFE',
              '50%': '#FFFFFF',
              '100%': '#FFCA5C',
            }}
            style={{ width: trackWidth, height: size }}
          >
            <Rect x={0} y={0} width={trackWidth} height={size} />
          </LinearGradient>
        </View>
        <View
          style={[
            styles.text,
            {
              transform: [
                {
                  translateX: -size / 2,
                },
                {
                  translateY: -size / 2,
                },
                {
                  rotate: '-90deg',
                },
              ],
              height: lineHeight,
              width: size,
              left: trackWidth + lineHeight / 2,
              bottom: -lineHeight / 2,
            },
          ]}
        >
          <TYText
            style={[{ fontSize: 12 }, textStyle]}
            ref={(ref: TYText) => {
              this.textRef = ref;
            }}
          >
            {label}: {this.formatPercent(temp)}%
          </TYText>
        </View>

        <Animated.View
          style={{
            position: 'absolute',
            right: trackWidth + 4,
            top: Animated.add(
              -8,
              this.animation.interpolate({
                inputRange: [min, max],
                outputRange: [size, 0],
              })
            ),
          }}
        >
          <Svg width="22" height="16">
            <Path
              ref={(ref: Path) => {
                this.pathRef = ref;
              }}
              d="M2,2L17,2L20,8L17,14L2,14Z"
              stroke="#fff"
              strokeWidth={2}
              fill={color}
              strokeLinecap="round"
              strokeLinejoin="round"
            />
          </Svg>
        </Animated.View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  text: {
    alignItems: 'flex-start',
    bottom: 0,
    justifyContent: 'center',
    position: 'absolute',
  },
});
