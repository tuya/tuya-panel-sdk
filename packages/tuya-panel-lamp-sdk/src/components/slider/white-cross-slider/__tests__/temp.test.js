import React from 'react';
import { shallow } from 'enzyme';
import Temp from '../Temp';

describe('Temp components', () => {
  it('basic render', () => {
    const wrapper = shallow(<Temp temp={30} min={0} max={1000} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('componentWillReceiveProps render', () => {
    let temp = 0;
    let size = 10;

    const wrapper = shallow(<Temp temp={temp} size={size} min={0} max={1000} />);
    expect(wrapper).toMatchSnapshot();

    temp = 30;
    wrapper.setProps({ temp });

    size = 30;
    wrapper.setProps({ size });

    wrapper.instance().update(500);

    wrapper.instance().handleUpdateThumb({ value: 500 });

    wrapper.setProps({ formatPercent: v => v });

    wrapper.unmount();
  });
});
