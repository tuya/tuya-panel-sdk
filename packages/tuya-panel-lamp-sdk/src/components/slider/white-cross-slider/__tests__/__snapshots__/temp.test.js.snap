// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Temp components basic render 1`] = `
<View
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": NaN,
        "height": undefined,
        "overflow": "hidden",
        "width": undefined,
      }
    }
  >
    <LinearGradient
      gradientId="white-cross-temp"
      stops={
        Object {
          "0%": "#CDECFE",
          "100%": "#FFCA5C",
          "50%": "#FFFFFF",
        }
      }
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
      x1="0%"
      x2="0%"
      y1="0%"
      y2="100%"
    >
      <Rect
        height={0}
        rx={0}
        ry={0}
        width={0}
        x={0}
        y={0}
      />
    </LinearGradient>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "bottom": 0,
          "justifyContent": "center",
          "position": "absolute",
        },
        Object {
          "bottom": -9,
          "height": 18,
          "left": NaN,
          "transform": Array [
            Object {
              "translateX": NaN,
            },
            Object {
              "translateY": NaN,
            },
            Object {
              "rotate": "-90deg",
            },
          ],
          "width": undefined,
        },
      ]
    }
  >
    <TYText
      align={null}
      color={null}
      size={null}
      style={
        Array [
          Object {
            "fontSize": 12,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
      type={null}
      weight={null}
    >
      : 
      3
      %
    </TYText>
  </View>
  <AnimatedComponent
    style={
      Object {
        "position": "absolute",
        "right": NaN,
        "top": NaN,
      }
    }
  >
    <Svg
      height="16"
      preserveAspectRatio="xMidYMid meet"
      width="22"
    >
      <Path
        d="M2,2L17,2L20,8L17,14L2,14Z"
        fill="rgb(255, 210, 168)"
        stroke="#fff"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </Svg>
  </AnimatedComponent>
</View>
`;

exports[`Temp components componentWillReceiveProps render 1`] = `
<View
  style={
    Object {
      "height": 10,
      "width": undefined,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": NaN,
        "height": 10,
        "overflow": "hidden",
        "width": undefined,
      }
    }
  >
    <LinearGradient
      gradientId="white-cross-temp"
      stops={
        Object {
          "0%": "#CDECFE",
          "100%": "#FFCA5C",
          "50%": "#FFFFFF",
        }
      }
      style={
        Object {
          "height": 10,
          "width": undefined,
        }
      }
      x1="0%"
      x2="0%"
      y1="0%"
      y2="100%"
    >
      <Rect
        height={10}
        rx={0}
        ry={0}
        width={0}
        x={0}
        y={0}
      />
    </LinearGradient>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "bottom": 0,
          "justifyContent": "center",
          "position": "absolute",
        },
        Object {
          "bottom": -9,
          "height": 18,
          "left": NaN,
          "transform": Array [
            Object {
              "translateX": -5,
            },
            Object {
              "translateY": -5,
            },
            Object {
              "rotate": "-90deg",
            },
          ],
          "width": 10,
        },
      ]
    }
  >
    <TYText
      align={null}
      color={null}
      size={null}
      style={
        Array [
          Object {
            "fontSize": 12,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
      type={null}
      weight={null}
    >
      : 
      0
      %
    </TYText>
  </View>
  <AnimatedComponent
    style={
      Object {
        "position": "absolute",
        "right": NaN,
        "top": 2,
      }
    }
  >
    <Svg
      height="16"
      preserveAspectRatio="xMidYMid meet"
      width="22"
    >
      <Path
        d="M2,2L17,2L20,8L17,14L2,14Z"
        fill="rgb(255, 208, 163)"
        stroke="#fff"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </Svg>
  </AnimatedComponent>
</View>
`;
